enhancements/cluster-logging/content-filter.md

```diff
@@ -0,0 +1,212 @@
+---
+title: content-filter
+authors:
+  - "@alanconway"
+reviewers: 
+  - "@jcantrill"
+approvers:
+  - "@jcantrill"
+api-approvers: 
+  - "@jcantrill"
+creation-date: 2023-11-03
+last-updated:  2023-11-03
+tracking-link:
+  - https://issues.redhat.com/browse/LOG-2155
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Content filters for log records
+
+## Summary
+
+Allow users to reduce the volume of log data by:
+1. Dropping unwanted log records completely.
+2. Pruning unwanted fields from log records.
+
+The new prune/drop content filters use the same framework as the kube-api-audit filter.
+This framework can be extended with new types of filters in future.
+
+**NOTE**: Content filters are distinct from input selectors.
+Input selectors select or ignore entire log _streams_ based on _source metadata_.
+Content filters _edit_ log streams (remove and modify records) based on _record content_.
+
+## Motivation
+
+Collecting all logs from a cluster produces a large amount of data, which can be expensive to transport and store.
+A lot of log data is low-value noise that does not need to be stored.
+
+The ideal solution is to configure each application to generate only valuable log data, however:
+- Not all applications can be tuned to remove noise while keeping important data.
+- Configuration of applications is not always under the control of the cluster administrator.
+- The value of logs varies depending on use: for example debugging a specific problem vs. monitoring overall health.
+
+### User Stories
+
+* As a logging administrator I want to drop log records by severity or other field values in the record.
+* As a logging administrator I want to prune log records by removing pod annotations, or other uninteresting fields.
+
+### Goals
+
+The user can configure the ClusterLogForwarder to drop and prune log records according to their requirements.
+
+### Non-Goals
+
+No complex record transformations, for example adding new fields with computed values.
+More complex filters MAY be added in future, but not as part of this enhancement.
+
+## Proposal
+
+### Workflow Description
+
+**log administrator** is a human responsible for configuring log collection, storage and forwarding in a cluster.
+
+1. The log administrator creates named `filter` sections in a `ClusterLogForwarder`
+2. Pipelines that require filtering have a `filterRef` field with a list of filter names.
+3. The collector edits log streams according to the filters before forwarding.
+
+### API Extensions
+
+A new `filters` section in the `ClusterLogForwarder` allows named filters to be defined.
+This proposal defines two types of filter "prune" and "drop".
+
+#### Prune filters
+
+A "prune" filter removes fields from each record passing through the filter.
+
+##### API
+
+``` yaml
+spec:
+  filters:
+  - name: ""        # User selects filter name
+    type: prune
+    prune:
+      - matches: ""    # Regular expression, remove all fields with paths that match.
```

Comment:
I like the idea of regex but is these technically feasible with vector VRL?  These are paths but when implementing the filter we will be dealing with an object structure.  It seems like this will be difficult when spanning the "dots" in a path

Reply:
Understood. Mitigating features:
- it's not hard to correctly quote dots with '\\' but it is a pain since there are so many dots.
- a regexp like "kubernetes.namespace_name" will match as expected. It's not 100% correct in that it would also match "kubernetesXnamespace_name" but odds are it will work as desired.
I'm happy to restore "equal" and "notEqual" if you feel it's important, I'm in two minds myself. I do like the simplicity of a single "match" operator, but if it is a pain for the majority of users then restoring "equal" might be better.

Reply:
Agreed, returned to in/notIn